Refactor the course search to  a combined event and course search

nikomakela opened the pull request:
LIIKUNTA-339.
A new combined search is there to take the place of the AdvancedSearch component. The CombinedSearch actually uses the AdvancedSearch as a subcomponent, but also enriches it with new search utilities component, that currently includes
the search tabs, that can be used to change the search type. Also a lazy query is done after the primary data is loaded, just to fetch the count of the result items in the secondary search.

<img width="991" alt="image" src="https://user-images.githubusercontent.com/389204/210264453-e8f525ef-a2c8-4c4a-b352-5120ffab34ba.png">


[LIIKUNTA-339]: https://helsinkisolutionoffice.atlassian.net/browse/LIIKUNTA-339?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nikomakela committed:
chore: improve the aria label of other event time's show-button

nikomakela committed:
test: add a missing navigation context provider

nikomakela committed:
fix: use effect for similar events handler for loaded events

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr133.dev.hel.ninja** :rocket::rocket::rocket:

nikomakela committed:
chore(sports): remove the needless search filters

LIIKUNTA-339. Remove the needless search filters copied from Hobbies.
Fix the tests and their mocking related to the searches.
Removed the needless footer categories.
Fix the mocks of other event times in all apps.
test: fix to event search page container
fix: typecheck issues

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr133.dev.hel.ninja** :rocket::rocket::rocket:

nikomakela commented:
It should be checked that the results are narrowed down to the Sports Events. In the demo session it was noticed that the result set contains items that are not related to sport, which we do not want. I might have deleted too many filters from the search filters.

- [ ] Check that the sport filter is working

nikomakela commented:
During the demo it was tested that filtering with a text breaks the query.

- [ ] Fix the combined search with filters on
- [ ] Test that the tabs works as expected when more filtering is added

nikomakela committed:
feat(sports): search utilities to change the event type of search

LIIKUNTA-339.
A new combined search is there to take the place of the AdvancedSearch component.
The CombinedSearch actually uses the AdvancedSearch as a subcomponent, but
also enriches it with new search utilities component, that currently includes
the search tabs, that can be used to change the search type.
Also a lazy query is done after the primary data is loaded, just to fetch the
count of the result items in the secondary search.
fix(sports): query with filters and event type

LIIKUNTA-339.
The query was broken when an eventType was selected
and a filter like text-param was added. A result for this was an empty white page
and a query done only for the secondary search.
Now the tab can be changed after the search is done with a newly added filter
and also both the tabs are refreshed when the filter is added.

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr133.dev.hel.ninja** :rocket::rocket::rocket:

nikomakela committed:
fix(sports): go back button routing from the venue or event details page

LIIKUNTA-339

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr133.dev.hel.ninja** :rocket::rocket::rocket:

melniiv commented on apps/sports-helsinki/src/domain/search/combinedSearch/CombinedSearch.tsx:
```diff
@@ -0,0 +1,95 @@
+    ? (params.eventType as EventTypeId)
+    : null;
+
+  const switchToGeneralEventSearch = () => {
```
pretty much the same implementation as method below, could it be some generic used with eventtypeid param?

melniiv commented on apps/sports-helsinki/src/domain/search/combinedSearch/searchUtilities/searchTabs.module.scss:
```diff
@@ -0,0 +1,35 @@
+    color: var(--color-black-80);
+  }
+
+  .secondaryButton {
```
i ll refactor that in styling tabs ticket

melniiv commented on apps/sports-helsinki/src/domain/search/eventSearch/SearchPage.tsx:
```diff
@@ -18,27 +20,34 @@ import { toast } from 'react-toastify';
+  const primarySearchEventType = eventTypes.find(
+    (e) => (params.eventType as EventTypeId) === e
+  ) as EventTypeId;
+  const secondarySearchEventType = eventTypes.find(
```
looks like it works only when there are only 2 types in array, which is our case:)

nikomakela replied:
True :) I thought about that too, but didn't want to overcomplicate it.

nenonja commented on apps/sports-helsinki/src/domain/search/combinedSearch/CombinedSearch.tsx:
```diff
@@ -0,0 +1,95 @@
+  searchStatus: SearchResultItemsCount;
+} & AdvancedSearchProps;
+
+const CombinedSearch: React.FC<CombinedSearchProps> = ({
```
I think this component would benefit the compounded components pattern as has kinda compounded/combined nature: 
https://blog.logrocket.com/understanding-react-compound-components/
With the pattern it could look something like:
```
<CombinedSearch>
   <AdvancedSearch ... />
   <Tabs>
      <SearchTab .../>
      <SearchTab .../>
    </Tabs>
    {/* other utilities like a venues map search switcher */}
<CombinedSearch>    
```
It might be easier to maintain and re-use?

nikomakela replied:
Sound interesting and this could definitely be tried. I just didn't fully understand the concept yet. I need to read more.

nikomakela replied:
I'll continue doing this in a new PR

nenonja commented on apps/sports-helsinki/src/utils/routerUtils.ts:
```diff
@@ -82,17 +82,18 @@ export const removeQueryParamsFromRouter = (
 
   if (removeList.length > 0) {
-    removeList.forEach((param) => delete router.query[param]);
+    removeList.forEach((param) => delete queryObject[param]);
```
the method name is `removeQueryParamsFromRouter` but now only cloned `queryObject` is changed but not the original query object? Was that on purpose?

nikomakela replied:
I actually don't know what has been the original purpose of this, but earlier it did not work and now it does 😄 . I think the old command was done for react-router and not for next-router and therefore it worked differently. The thing was that nothing was actually done to queryObject that was used in the next router push-command.

nenonja reviewed (commented):
Looks good! Some questions, comments, and refactoring ideas... but all can be ignored or done later

nikomakela commented on apps/events-helsinki/src/domain/event/__tests__/EventPageContainer.test.tsx:
```diff
@@ -69,7 +69,7 @@ const otherEventsRequest = {
   variables: {
     include: ['in_language', 'keywords', 'location', 'audience'],
-    sort: 'start_time',
+    sort: 'end_time',
```
yea, I investigated this too. The thing is that a apollo error was raised but ignored in the tests. So the test was kind of "broken".

nikomakela committed:
chore: remove needless linter check ignore commands and unused getter

nikomakela commented on apps/events-helsinki/src/domain/event/similarEvents/SimilarEvents.tsx:
```diff
@@ -61,9 +61,11 @@ const SimilarEvents: React.FC<Props> = ({
-  if (hasCards) {
-    onEventsLoaded && onEventsLoaded(cards.length);
-  }
+  React.useEffect(() => {
```
Yep, the unit tests threw some React-error

nikomakela commented on apps/sports-helsinki/src/domain/app/AppConfig.ts:
```diff
@@ -47,8 +47,12 @@ class AppConfig {
     return process.env.NEXT_PUBLIC_CMS_PAGES_CONTEXT_PATH ?? '/pages';
   }
 
+  /** @deprecated since the event search has a splitted view
```
Yep. I'll remove it! It was only used in a AppConfig tests anymore.

nikomakela commented on apps/hobbies-helsinki/src/domain/event/__tests__/EventPageContainer.test.tsx:
```diff
@@ -149,7 +153,9 @@ it('should render info and load other events + similar events', async () => {
 
   // click show other times
   // eslint-disable-next-line testing-library/no-unnecessary-act
```
It could be, but the EventPageContainer-component is still implemented inside the application folder, so I would say that an unit test needs to be where the actual component is.

nikomakela commented on apps/hobbies-helsinki/config/jest/TestProviders.tsx:
```diff
@@ -47,9 +50,11 @@ function TestProviders({ mocks, children, router }: Props) {
-            <RouterContext.Provider value={{ ...router, ...mockRouter }}>
-              {children}
-            </RouterContext.Provider>
+            <NavigationContext.Provider value={{}}>
```
Yes, I've tried to move the TestProvider couple of times already, but there are quite many imports from the application, which would not be available in the common-test-package.

```
import EventsConfigProvider from '../../src/common-events/configProvider/ConfigProvider';
import eventsDefaultConfig from '../../src/common-events/configProvider/defaultConfig';
import { ROUTES } from '../../src/constants';
import { getLocalizedCmsItemUrl } from '../../src/utils/routerUtils';
import { initI18n as i18n } from './initI18n';

const CMS_API_DOMAIN = 'tapahtumat.cms.test.domain.com';
```

Therefore, a lot of refactoring would be needed and not much would be included in the common part. Could be done at some point, but I don't see much advantage of it since most of the content comes from the application specific configuration.

nenonja replied:
Maybe this could be done at some point later

nikomakela commented on apps/hobbies-helsinki/src/domain/event/eventInfo/OtherEventTimes.tsx:
```diff
@@ -17,7 +17,7 @@ const EVENTS_LIST_LIMIT = 3;
 
 const OtherEventTimes: React.FC<{ event: EventFields }> = ({ event }) => {
-  const { t } = useTranslation('event');
+  const { t } = useEventTranslation();
```
OtherEventTimes imports the useOtherEventTimes from the application specific queryUtils:

```
import { useOtherEventTimes } from '../queryUtils';
```

That function uses `useOtherEventTimesVariables` which uses a lot of application specific stuff.

Therefore it seems like we cannot move the OtherEventTimes very easily 😞

nenonja replied:
Maybe this could be done at some point later

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr133.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr133.dev.hel.ninja** :rocket::rocket::rocket:
